src/box/txn_limbo.c

```diff
@@ -1086,6 +1091,84 @@ txn_limbo_filter_request(struct txn_limbo *limbo,
 	}
 }
 
+static uint32_t sync_system_space_ids[] = {
+	/* BOX_VINYL_DEFERRED_DELETE_ID, local space*/
+	BOX_SCHEMA_ID,
+	BOX_COLLATION_ID,
+	BOX_VCOLLATION_ID,
+	BOX_SPACE_ID,
+	BOX_VSPACE_ID,
+	BOX_SEQUENCE_ID,
+	/* BOX_SEQUENCE_DATA_ID, synchronized by synchronous space operations */
+	BOX_VSEQUENCE_ID,
+	BOX_INDEX_ID,
+	BOX_VINDEX_ID,
+	BOX_FUNC_ID,
+	BOX_VFUNC_ID,
+	BOX_USER_ID,
+	BOX_VUSER_ID,
+	BOX_PRIV_ID,
+	BOX_VPRIV_ID,
+	BOX_CLUSTER_ID,
+	BOX_TRIGGER_ID,
+	BOX_TRUNCATE_ID,
+	BOX_SPACE_SEQUENCE_ID,
+	BOX_VSPACE_SEQUENCE_ID,
+	BOX_FK_CONSTRAINT_ID,
+	BOX_CK_CONSTRAINT_ID,
+	BOX_FUNC_INDEX_ID,
+	BOX_SESSION_SETTINGS_ID,
+};
+
+/**
+ * Controls whether to consider system spaces indefinitely synchronous when the
+ * synchronous queue is claimed.
+ */
+bool box_consider_system_spaces_synchronous;
+TWEAK_BOOL(box_consider_system_spaces_synchronous);
+
+/**
+ * Update the state of synchronous replication for system spaces in the
+ * `sync_system_space_ids` list.
+ */
+static void
+txn_limbo_update_sync_system_spaces(struct txn_limbo *limbo, uint16_t req_type,
+				    bool is_rollback)
+{
+	assert(req_type == IPROTO_RAFT_PROMOTE ||
+	       req_type == IPROTO_RAFT_DEMOTE);
+	if (is_rollback)
+		req_type = req_type == IPROTO_RAFT_PROMOTE ?
+			   IPROTO_RAFT_DEMOTE : IPROTO_RAFT_PROMOTE;
+	if (recovery_state <= INITIAL_RECOVERY ||
+	    (!is_rollback &&
+	     ((req_type == IPROTO_RAFT_PROMOTE &&
+	       limbo->owner_id != REPLICA_ID_NIL) ||
+	      (req_type == IPROTO_RAFT_DEMOTE &&
+	       limbo->owner_id == REPLICA_ID_NIL))))
+		return;
+	for (size_t i = 0; i < lengthof(sync_system_space_ids); ++i) {
+		uint32_t space_id = sync_system_space_ids[i];
+		struct space *space = space_by_id(space_id);
+		/*
+		 * Some system spaces can be absent in an older version of the
+		 * schema.
+		 */
+		if (space == NULL)
+			continue;
+		assert(space != NULL);
+		space->is_sync = space->def->opts.is_sync ||
+				 (box_consider_system_spaces_synchronous &&
+				  req_type == IPROTO_RAFT_PROMOTE);
+		assert(space->lua_ref != LUA_NOREF);
+		lua_rawgeti(tarantool_L, LUA_REGISTRYINDEX, space->lua_ref);
+		lua_getfield(tarantool_L, -1, "state");
+		lua_pushboolean(tarantool_L, space->is_sync);
+		lua_setfield(tarantool_L, -2, "is_sync");
+		lua_pop(tarantool_L, 2);
+	}
```

Comment:
I think the list of spaces (or rather a blacklist) and the code manipulating their is_sync should be in `space.c` (or `schema.c` or somewhere near there). This stuff is internal to the space subsystem.

You could have two methods:
`system_spaces_make_sync`
and
`system_spaces_make_async`.

Limbo would call them depending on the request being processed (`PROMOTE` or `DEMOTE`). And sometimes they would take no action, if `consider_system_spaces_synchronous` is `false`.

Besides, you could call these methods when the user changes `consider_system_spaces_synchronous` on the fly, without promoting or demoting anyone.

Reply:
> I think the list of spaces (or rather a blacklist) and the code manipulating their is_sync should be in space.c (or schema.c or somewhere near there). This stuff is internal to the space subsystem.

Moved to `box/schema.{h, cc}`.

> Besides, you could call these methods when the user changes consider_system_spaces_synchronous on the fly, without promoting or demoting anyone

How should I call them if do not know what state should I transition the system space to? 

> when the user changes consider_system_spaces_synchronous on the fly

I don't think this is a valid use case.

Reply:
> Moved to box/schema.{h, cc}.

@CuriousGeorgiy they're still in `txn_limbo`

Reply:
Sorry, I forgot to push the fixes.